apps/api/src/Match/Match.service.ts

```diff
@@ -0,0 +1,110 @@
+import { Injectable, BadRequestException } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { TransformUserService } from 'src/TransformUser/TransformUser.service';
+import { Repository } from 'typeorm';
+import { Match, User, Userfront } from 'types';
+import { UsersService } from '../users/users.service';
+
+export interface MatchFront {
+  id: string;
+  score_winner: number;
+  score_looser: number;
+  looser: Userfront | null;
+  winner: Userfront | null;
+}
+
+@Injectable()
+export class MatchService {
+  constructor(
+    @InjectRepository(Match)
+    private readonly matchRepository: Repository<Match>,
+    @InjectRepository(User)
+    private readonly userRepository: Repository<User>,
+    private readonly userService: UsersService,
+    private readonly transformUserService: TransformUserService,
+  ) {}
+
+  #getXp(
+    winner: User,
+    looser: User,
+    score_winner: number,
+    score_looser: number,
+  ): number {
+    if (winner && looser && score_looser && score_winner) return 5;
+    //calcul xp selon niveau et score
+    return 5;
+  }
+
+  async addMatch(
+    winner_id: string,
+    looser_id: string,
+    score_winner: number,
+    score_looser: number,
+  ): Promise<number> {
+    const winner = await this.userRepository.findOneBy({ id: winner_id });
+    if (!winner) throw new BadRequestException('winner not find');
+
+    const looser = await this.userRepository.findOneBy({ id: looser_id });
+    if (!looser) throw new BadRequestException('looser not find');
+
+    const newMatch = new Match();
+    newMatch.winner_id = winner;
+    newMatch.looser_id = looser;
+    newMatch.score_winner = score_winner;
+    newMatch.score_looser = score_looser;
+    await this.matchRepository.save(newMatch);
+
+    this.userService.updateLevel(
+      winner_id,
+      this.#getXp(winner, looser, score_winner, score_looser),
+    );
+    return 1;
+  }
+
+  async getMatch(user_id: string): Promise<MatchFront[]> {
+    const user = await this.userRepository.findOne({
+      relations: {
+        win: { looser_id: true , winner_id: true},
+
+        defeat: { winner_id: true, looser_id: true },
+      },
+      where: {
+        id: user_id,
+      },
+    });
+    if (!user) throw new BadRequestException('user not found');
+
+    const winArray: MatchFront[] = [];
+
+    const looseArray: MatchFront[] = [];
+
+    for (let i = 0; i < user.win.length; i++) {
```

Comment:
I believe
`let winArray = user.win.map((el) => return ({id: el.id, score_winner: el.score_winner, score_looser: el.score_looser, looser: await this.transformUserService.transform(el.looser_id), winner: null}))`
would be more appropriate

Reply:
ben j'ai un probleme avec les map en async ... ca ne marche pas ... c'est pour ca que je fais des boucles partout. tu as peut etre une explication.

Reply:
ah oui, d'ailleurs j'aurais du la declarer en async dans mon exemple, ca devrait etre:
`const winArray = user.win.map(_**async**_ (el) => return ({id: el.id, score_winner: el.score_winner, score_looser: el.score_looser, looser: await this.transformUserService.transform(el.looser_id), winner: null}))`
Sinon je ne vois pas trop